### PR TITLE
Screenshot: Add GUI screenshot tool

### PR DESCRIPTION
**Pull request by xslendix**
This is a basic GUI tool that allows you to take screenshots inside Serenity.

To-do:

- [x] Basic UI/UX design
- [x] Full screen screenshots
- [x] Custom screenshots
- [x] Delay option
- [x] Copy to clipboard checkbox
- [x] Setting autosave
- [x] Super+S shortcut for quick opening

### CODE REVIEW

File: Userland/Applications/CMakeLists.txt [hunk 1]
```diff
@@ -33,3 +33,4 @@ add_subdirectory(TextEditor)
 add_subdirectory(ThemeEditor)
 add_subdirectory(VideoPlayer)
 add_subdirectory(Welcome)
+add_subdirectory(Screenshoter)
```
Comment (linusg) [inline Userland/Applications/CMakeLists.txt:36]: this list is clearly sorted, please keep it that way

File: Userland/Applications/Screenshoter/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+compile_gml(ScreenshoterWindow.gml ScreenshoterWindowGML.h screenshoter_window_gml)
+
+set(SOURCES
+    ScreenshoterWindowGML.h
+    ScreenshoterWidget.cpp
+    ScreenshoterWidget.h
+    main.cpp
+)
+
+serenity_app(Screenshoter ICON app-welcome)
```
Comment (linusg) [inline Userland/Applications/Screenshoter/CMakeLists.txt:15]: why is this using the welcome app icon?

File: Userland/Applications/Screenshoter/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+target_link_libraries(Screenshoter LibGUI LibWeb)
```
Comment (linusg) [inline Userland/Applications/Screenshoter/CMakeLists.txt:16]: why is this linking with LibWeb?

File: Userland/Applications/Screenshoter/ScreenshoterWidget.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    SelectableLayover(GUI::Window* window)
+        : m_window(window)
+        , m_background_color(palette().threed_highlight().with_alpha(128))
+    {
+        set_override_cursor(Gfx::StandardCursor::Crosshair);
+    }
+
+    virtual ~SelectableLayover() override {};
+
+    Gfx::IntRect region() const
```
Comment (linusg) [inline Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:40]:
```suggestion
    Gfx::IntRect const& region() const
```

File: Userland/Applications/Screenshoter/ScreenshoterWidget.cpp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    m_delay_spinbox = *find_descendant_of_type_named<GUI::SpinBox>("delay");
+    m_cancel_button = *find_descendant_of_type_named<GUI::Button>("cancel_button");
+    m_ok_button = *find_descendant_of_type_named<GUI::Button>("ok_button");
+
+    m_cancel_button->on_click = [&](auto) {
+        GUI::Application::the()->quit();
+    };
+
+    m_ok_button->on_click = [&](auto) {
+        String output_path = Core::DateTime::now().to_string(String::formatted("/home/{}/screenshot-%Y-%m-%d-%H-%M-%S.png", getlogin()));
```
Comment (linusg) [inline Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:108]: This should use `Core::StandardPaths::home_directory()`

File: Userland/Applications/Screenshoter/ScreenshoterWidget.cpp [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            bitmap = bitmap->cropped(crop_region);
+        }
+
+        auto encoded_bitmap = Gfx::PNGWriter::encode(*bitmap);
+        if (encoded_bitmap.is_empty()) {
+            warnln("Failed to encode PNG");
+            return;
+        }
+
+        auto file_or_error = Core::File::open(output_path, Core::OpenMode::ReadWrite);
```
Comment (linusg) [inline Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:148]: We don't need to read from this file so `OpenMode::WriteOnly` is sufficient

File: Userland/Applications/Screenshoter/ScreenshoterWidget.cpp [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        if (!file.write(encoded_bitmap.data(), encoded_bitmap.size())) {
+            warnln("Failed to write PNG");
+            return;
+        }
+    
+        auto notification = GUI::Notification::construct();
+        notification->set_title("Screenshot taken");
+        notification->set_text(String::formatted("Screenshot saved at {}", output_path));
+        notification->show();
+
```
Comment (linusg) [inline Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:164]:
```suggestion
```

File: Userland/Applications/Screenshoter/ScreenshoterWindow.gml [hunk 8]
```diff
@@ -0,0 +1,78 @@
+@GUI::Widget {
+    fill_with_background_color: true
+    layout: @GUI::HorizontalBoxLayout {
+        margins: [8,8,8,8]
```
Comment (linusg) [inline Userland/Applications/Screenshoter/ScreenshoterWindow.gml:4]:
```suggestion
        margins: [8, 8, 8, 8]
```

File: Userland/Applications/Screenshoter/ScreenshoterWindow.gml [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+        @GUI::RadioButton {
+            name: "wholescreen"
+            text: "Whole screen"
+            checked: true
+        }
+        @GUI::RadioButton {
+            name: "activewindow"
+            text: "Active window"
+        }
```
Comment (linusg) [inline Userland/Applications/Screenshoter/ScreenshoterWindow.gml:24]:
```suggestion

        @GUI::RadioButton {
            name: "activewindow"
            text: "Active window"
        }

```

File: Userland/Applications/Screenshoter/main.cpp [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+    if (unveil("/tmp/portal/notify", "rw") < 0) {
+        perror("unveil");
+        return 1;
+    }
+
+    if (unveil("/bin/Screenshoter", "x") < 0) {
+        perror("unveil");
+        return 1;
+    }
```
Comment (linusg) [inline Userland/Applications/Screenshoter/main.cpp:40]: Why is this needed?

File: Userland/Applications/Screenshoter/main.cpp [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    if (unveil(nullptr, nullptr) < 0) {
+        perror("unveil");
+        return 1;
+    }
+
+    // auto app_icon = GUI::Icon::default_icon("app-welcome");
```
Comment (linusg) [inline Userland/Applications/Screenshoter/main.cpp:52]: Please don't include commented-out code

**Review by linusg — CHANGES_REQUESTED** [review]
Added -> Add in the commit message

File: Userland/Applications/Screenshoter/ScreenshoterWidget.cpp (hunk 4 above)
Comment (xslendix) [reply Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:42]: This will break the build.

File: Userland/Applications/CMakeLists.txt [hunk 12]
```diff
@@ -24,6 +24,7 @@ add_subdirectory(Piano)
 add_subdirectory(PixelPaint)
 add_subdirectory(Run)
 add_subdirectory(Settings)
+add_subdirectory(Screenshoter)
```
Comment (linusg) [inline Userland/Applications/CMakeLists.txt:27]: This should go before `Settings`

File: Userland/Applications/Screenshoter/ScreenshoterWidget.cpp [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    m_cancel_button->on_click = [&](auto) {
+        GUI::Application::the()->quit();
+    };
+
+    m_ok_button->on_click = [&](auto) {
+        String output_path = Core::DateTime::now().to_string(String::formatted("{}/screenshot-%Y-%m-%d-%H-%M-%S.png", Core::StandardPaths::home_directory()));
+
+        GUI::Application::the()->active_window()->close();
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(m_delay_spinbox->value()));
```
Comment (linusg) [inline Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:114]: We don't use the C++ standard library.

Comment (xslendix) [reply Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:114]: How else would I sleep the current thread?

Comment (xslendix) [reply Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:114]: I can't see any other way of sleeping for x amount of milliseconds.

Comment (doctor-rd) [reply Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:114]: May be you can use LibCore/Timer.h.

File: Userland/Applications/Screenshoter/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,16 @@
+serenity_component(
+    Screenshoter
```
Comment (awesomekling) [inline Userland/Applications/Screenshoter/CMakeLists.txt:2]: Please call this application "Screenshot".

File: Userland/Applications/Screenshoter/ScreenshoterWidget.cpp [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+    m_copy_checkbox = *find_descendant_of_type_named<GUI::CheckBox>("copy_to_clipboard");
+    m_cancel_button = *find_descendant_of_type_named<GUI::Button>("cancel_button");
+    m_ok_button = *find_descendant_of_type_named<GUI::Button>("ok_button");
+
+    m_cancel_button->on_click = [&](auto) {
+        GUI::Application::the()->quit();
+    };
+
+    m_ok_button->on_click = [&](auto) {
+        GUI::Application::the()->active_window()->close();
```
Comment (awesomekling) [inline Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:111]: We don't need to ask `Application` what the active window is, it's the window we are in:
```cpp
window()->close();
```

File: Userland/Applications/Screenshoter/ScreenshoterWidget.cpp [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+
+        String output_path = Core::DateTime::now().to_string(String::formatted("{}/screenshot-%Y-%m-%d-%H-%M-%S.png", Core::StandardPaths::home_directory()));
```
Comment (awesomekling) [inline Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:113]: I don't think we should hard-code the home directory here. Instead, put the screenshot in the current directory. Where you run the app from is up to you.

File: Userland/Applications/Screenshoter/ScreenshoterWidget.cpp [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+    Gfx::IntRect crop_region;
+    if (selected) {
+        auto window = GUI::Window::construct();
+        auto& container = window->set_main_widget<SelectableLayover>(window);
+
+        window->set_title("Screenshoter");
+        window->set_has_alpha_channel(true);
+        window->set_fullscreen(true);
+        window->show();
+        GUI::Application::the()->exec();
```
Comment (awesomekling) [inline Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:133]: It is not appropriate to re-enter `Application::exec()` like this when it's already on the stack.
Instead we should return from here and give back control to the event loop that way.
The window can be kept alive by making it global.

File: Userland/Applications/Screenshoter/ScreenshoterWidget.cpp [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+
+        String output_path = Core::DateTime::now().to_string(String::formatted("{}/screenshot-%Y-%m-%d-%H-%M-%S.png", Core::StandardPaths::home_directory()));
+
+        usleep(m_delay_spinbox->value() * 1000);
+
+        save_screenshot(output_path, m_custom_button->is_checked(), m_copy_checkbox->is_checked());
+    };
+}
+
+void ScreenshoterWidget::save_screenshot(String output_path, bool selected = false, bool copy_to_clipboard = false)
```
Comment (awesomekling) [inline Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:121]: I don't see why this function needs default values for the `bool` parameters.

File: Userland/Applications/Screenshoter/ScreenshoterWindow.gml [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+        @GUI::Widget {
+            shrink_to_fit: true
+
+            layout: @GUI::HorizontalBoxLayout {
+            }
+
+            @GUI::Button {
+                name: "ok_button"
+                text: "OK"
+                fixed_height: 24
```
Comment (awesomekling) [inline Userland/Applications/Screenshoter/ScreenshoterWindow.gml:77]: Please don't override the default height of basic GUI buttons.

File: Userland/Applications/Screenshoter/ScreenshoterWindow.gml [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+            }
+            
+            @GUI::Button {
+                name: "cancel_button"
+                text: "Cancel"
+                fixed_height: 24
```
Comment (awesomekling) [inline Userland/Applications/Screenshoter/ScreenshoterWindow.gml:83]: Please don't override the default height of basic GUI buttons.

File: Userland/Applications/Screenshoter/main.cpp [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+#include "ScreenshoterWidget.h"
+#include <LibGUI/Application.h>
+#include <LibGUI/Icon.h>
+#include <LibGUI/Window.h>
+#include <unistd.h>
+
+int main(int argc, char** argv)
+{
+    if (pledge("stdio recvfd sendfd cpath wpath rpath unix proc exec", nullptr) < 0) {
```
Comment (awesomekling) [inline Userland/Applications/Screenshoter/main.cpp:15]: Why does this need `proc` and `exec`?

File: Userland/Applications/Screenshoter/ScreenshoterWidget.cpp [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+#include <LibGUI/RadioButton.h>
+#include <LibGUI/SpinBox.h>
+#include <LibGUI/CheckBox.h>
+#include <LibGUI/Label.h>
+#include <LibGUI/Notification.h>
+#include <LibGfx/PNGWriter.h>
+#include <LibGfx/Palette.h>
+#include <serenity.h>
+
+class SelectableLayover final : public GUI::Widget {
```
Comment (awesomekling) [inline Userland/Applications/Screenshoter/ScreenshoterWidget.cpp:30]: This name is really poor (layover != overlay).
Please call it something like `SelectionOverlay` instead.

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Userland/Applications/Screenshot/ScreenshotWidget.cpp [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+        notification->set_title("Screenshot taken");
+        notification->set_text("Screenshot saved to clipboard");
+        notification->show();
+
+        return;
+    }
+
+    if (selected) {
+        bitmap = bitmap->cropped(crop_region);
+    }
```
Comment (mustafaquraish) [inline Userland/Applications/Screenshot/ScreenshotWidget.cpp:187]: You're copying to the clipboard before cropping the image, even if there's a selection made. Instead of doing it this way, you should just pass in `crop_region` as the first argument for `GUI::WindowServerConnection::the().get_screen_bitmap`

File: Userland/Applications/Screenshoter/main.cpp [hunk 24]
```diff
@@ -34,6 +34,11 @@ int main(int argc, char** argv)
         return 1;
     }
```
Comment (doctor-rd) [inline Userland/Applications/Screenshoter/main.cpp]: This allows bypassing `Config::pledge_domains("Screenshoter");`.

**Comment by doctor-rd** [comment]
Unveiling `/tmp/portal/config` allows bypassing `Config::pledge_domains("Screenshoter");`.

File: Userland/Applications/Screenshot/ScreenshotWidget.cpp (hunk 23 above)
Comment (xslendix) [reply Userland/Applications/Screenshot/ScreenshotWidget.cpp:187]: I have done it this way since I was having some issues while testing. see ed3b1b0.

Comment (mustafaquraish) [reply Userland/Applications/Screenshot/ScreenshotWidget.cpp:118]: did you rebase to the new master? If you'll notice the PR that added this change also had a fix in the WindowServer code that passes you the screen bitmap to fix this issue.

Comment (xslendix) [reply Userland/Applications/Screenshot/ScreenshotWidget.cpp:118]: I did not, that explains why it didn't work for me, I will create a new commit to undo this change.

File: Userland/Applications/Screenshot/ScreenshotWidget.cpp [hunk 25]
```diff
@@ -113,9 +113,9 @@ void ScreenshotWidget::save_screenshot(String output_path, bool selected, bool c
         return;
     }
 
-    if (selected) {
-        bitmap = bitmap->cropped(crop_region);
-    }
+    // if (selected) {
+    //     bitmap = bitmap->cropped(crop_region);
+    // }
```
Comment (mustafaquraish) [inline Userland/Applications/Screenshot/ScreenshotWidget.cpp:118]: You should remove commented out code from the commits.

Comment (xslendix) [reply Userland/Applications/Screenshot/ScreenshotWidget.cpp:118]: Oops! Thought I removed that.

**Comment by trflynn89** [comment]
> Unveiling `/tmp/portal/config` allows bypassing `Config::pledge_domains("Screenshoter");`.

Wait, don't we prefer pledging specific domains?

File: Userland/Applications/Screenshot/ScreenshotWidget.cpp [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+
+        Config::write_string("Screenshot", "General", "ScreenshotType", new_type);
+        Config::write_i32("Screenshot", "General", "Delay", m_delay_spinbox->value());
+        Config::write_bool("Screenshot", "General", "CopyToClipboard", m_copy_checkbox->is_checked());
+
+        window()->close();
+
+        m_output_path = Core::DateTime::now().to_string("screenshot-%Y-%m-%d-%H-%M-%S.png");
+
+        usleep(m_delay_spinbox->value() * 1000);
```
Comment (trflynn89) [inline Userland/Applications/Screenshot/ScreenshotWidget.cpp:79]: As discussed on Discord, we should not be blocking the core event loop. Aside from just not being a nice thing to do, it will make the kernel / WindowServer think the application is unresponsive.

**Review by trflynn89 — CHANGES_REQUESTED** [review]

File: Userland/Applications/Screenshot/ScreenshotWidget.cpp (hunk 26 above)
Comment (xslendix) [reply Userland/Applications/Screenshot/ScreenshotWidget.cpp:79]: This has already been discussed on Discord. It should be perfectly fine to block it here since the main window gets closed beforehand. Also, using a Timer doesn't work. I might give it a try again tomorrow but for now it works perfectly and have never had any issues.

Comment (xslendix) [reply Userland/Applications/Screenshot/ScreenshotWidget.cpp:79]: I also tries pledging only the Screenshot domain but doesn't work as well since it still needs access to /tmp/portal/config. Don't ask me why though.

Comment (doctor-rd) [reply Userland/Applications/Screenshot/ScreenshotWidget.cpp:79]: You have to call pledge_domains before the very first call to unveil.

Comment (trflynn89) [reply Userland/Applications/Screenshot/ScreenshotWidget.cpp:79]: > Also, using a Timer doesn't work
It does - I posted a gist on Discord showing how to use it with this app

Comment (xslendix) [reply Userland/Applications/Screenshot/ScreenshotWidget.cpp:79]: Aha! Didn't see it. Imma try it out.

**Comment by xslendix** [comment]
CI complains because `LibConfig/Client.h` is put at the bottom of the includes. It is placed there due to a current bug with LibConfig where the build fails when placed any other place, to my knowledge.

File: Userland/Applications/Screenshot/ScreenshotWidget.cpp [hunk 27]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        Config::write_string("Screenshot", "General", "ScreenshotType", new_type);
+        Config::write_i32("Screenshot", "General", "Delay", m_delay_spinbox->value());
+        Config::write_bool("Screenshot", "General", "CopyToClipboard", m_copy_checkbox->is_checked());
+
+        m_output_path = Core::DateTime::now().to_string("screenshot-%Y-%m-%d-%H-%M-%S.png");
+
+        window()->set_frameless(true);
+        window()->set_opacity(0);
```
Comment (trflynn89) [inline Userland/Applications/Screenshot/ScreenshotWidget.cpp:87]: One thing that was pointed out on Discord after I posted that gist is that the better way to keep the app open once the window is closed is to use the `Application::set_quit_when_last_window_deleted()` API. Then, you should be able to do `window()->close();` here instead of this weird opacity thing I came up with :)

So something like:
```c++
GUI::Application::the()->set_quit_when_last_window_deleted(false);
window->close();
```

Comment (trflynn89) [reply Userland/Applications/Screenshot/ScreenshotWidget.cpp:87]: Then, in the timer callback, you should invoke this after saving the screenshot:
```c++
GUI::Application::the()->quit();
```

File: Userland/Applications/Screenshot/ScreenshotWidget.cpp [hunk 28]
```diff
@@ -0,0 +1,155 @@
...
+
+    m_window_sel = GUI::Window::construct();
+    m_overlay = m_window_sel->set_main_widget<SelectableOverlay>(m_window_sel);
+    m_overlay->callback = [&] {
+        save_screenshot(true);
+    };
+
+    m_window_sel->set_title("Screenshot");
+    m_window_sel->set_has_alpha_channel(true);
+    m_window_sel->set_fullscreen(true);
```
Comment (krkk) [inline Userland/Applications/Screenshot/ScreenshotWidget.cpp:56]: There’s no icon on the taskbar during selection.
Peek 2021-08-31 15-19

But maybe instead of adding an icon we could take the Assistant’s approach and hide the window from the taskbar on selection?
```cc
    m_window_sel->set_frameless(true);
```

File: Userland/Applications/Screenshot/ScreenshotWidget.cpp [hunk 29]
```diff
@@ -0,0 +1,155 @@
...
+        return;
+    }
+
+    if (m_copy_checkbox->is_checked()) {
+        GUI::Clipboard::the().set_bitmap(*bitmap);
+
+        auto notification = GUI::Notification::construct();
+        notification->set_title("Screenshot taken");
+        notification->set_text("Screenshot saved to clipboard");
+        notification->show();
```
Comment (krkk) [inline Userland/Applications/Screenshot/ScreenshotWidget.cpp:124]: An icon to an notification would be nice ;)

```cc
        notification->set_icon(GUI::Icon::default_icon("app-screenshot").bitmap_for_size(32));
```
proposed preview

File: Userland/Applications/Screenshot/ScreenshotWidget.cpp [hunk 30]
```diff
@@ -0,0 +1,155 @@
...
+    auto& file = *file_or_error.value();
+    if (!file.write(encoded_bitmap.data(), encoded_bitmap.size())) {
+        warnln("Failed to write PNG");
+        return;
+    }
+
+    auto notification = GUI::Notification::construct();
+    notification->set_title("Screenshot taken");
+    notification->set_text(String::formatted("Screenshot saved at {}", m_output_path));
+    notification->show();
```
Comment (krkk) [inline Userland/Applications/Screenshot/ScreenshotWidget.cpp:150]: same here

**Comment by doctor-rd** [comment]
You are still combining `unveil("/tmp/config","rw")` with `Config::pledge_domains`. For example in `TextEditor/main.cpp` you can see how to use `Config::pledge_domains`.

**Comment by xslendix** [comment]
How would I name the category in the commit message?
